pkg/kube/client/ssa/state_cache.go

```diff
@@ -0,0 +1,142 @@
+package ssa
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	xpresource "github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+// StateCacheManager lets you manage StateCache entries for XP managed
+// resources
+type StateCacheManager interface {
+	LoadOrNewForManaged(mg xpresource.Managed) StateCache
+	Remove(mg xpresource.Managed)
+}
+
+// StateCache is the interface for the caching a k8s
+// *unstructed.Unstructured object
+type StateCache interface {
+	SetState(state *unstructured.Unstructured, hash string)
+	GetState() (*unstructured.Unstructured, string)
+	HasState() bool
+}
+
+// DesiredStateCache is a concurrency-safe implementation of StateCache
+// that holds a cached k8s object state with a hash key of the associated
+// manifest.
+// Hash key can be used to determine the
+type DesiredStateCache struct {
+	logger    logging.Logger
+	mu        *sync.Mutex
+	extracted *unstructured.Unstructured
+	hash      string
+}
+
+// DesiredStateCacheOption lets you configure the DesiredStateCache parameters
+type DesiredStateCacheOption func(dsc *DesiredStateCache)
+
+// WithLogger sets the logger of DesiredStateCache.
+func WithLogger(l logging.Logger) DesiredStateCacheOption {
+	return func(w *DesiredStateCache) {
+		w.logger = l
+	}
+}
+
+// NewDesiredStateCache initializes a DesiredStateCache with given options
+func NewDesiredStateCache(opts ...DesiredStateCacheOption) *DesiredStateCache {
+	w := &DesiredStateCache{
+		logger: logging.NewNopLogger(),
+		mu:     &sync.Mutex{},
+	}
+	for _, f := range opts {
+		f(w)
+	}
+	return w
+}
+
+// GetState returns the stored desired state and the hash of associated
+// manifest
+func (dc *DesiredStateCache) GetState() (*unstructured.Unstructured, string) {
+	dc.mu.Lock()
+	defer dc.mu.Unlock()
+	return dc.extracted, dc.hash
+}
+
+// HasState returns whether the DesiredStateCache has a stored state
+func (dc *DesiredStateCache) HasState() bool {
+	dc.mu.Lock()
+	defer dc.mu.Unlock()
+	return dc.extracted != nil
+}
+
+// SetState stores the given desired k8s object state into
+// the DesiredStateCache
+func (dc *DesiredStateCache) SetState(state *unstructured.Unstructured, hash string) {
+	dc.mu.Lock()
+	defer dc.mu.Unlock()
+	dc.extracted = state
+	dc.hash = hash
+}
+
+// DesiredStateCacheStore stores the DesiredStateCache instances associated with the
+// managed resource instance.
+type DesiredStateCacheStore struct {
+	store  map[types.UID]*DesiredStateCache
+	logger logging.Logger
+	mu     *sync.Mutex
```

Comment:
nit: to indicate what mutex is protecting:

```suggestion
	logger logging.Logger
	
	mu     *sync.Mutex
	store  map[types.UID]*DesiredStateCache
```

Reply:
Should we use an `sync.RWMutex` and use `RLock()` when we only want to read?

Reply:
From pure performance perspective, we should do it. Each key here corresponds to an MR, writes happen once for each MR, and many reads.

Used regular Lock() here, because of the simpler code, compared to the efficiency gain.
We need LoadOrStore mechanics here for `DesiredStateCacheStore`, which we initialize per MR.
Using RLock() here, means double-checking the cache, which affects the simplicity of the code a little bit (in other PR reviews)

Anyways, see https://github.com/turkenh/provider-kubernetes/pull/1/commits/af48b4c061dabde41bd45d20b7234cb6b4705384